python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1380 @@
+    """
+
+    RunnerClass = MakeBrighterFatterKernelTaskRunner
+    ConfigClass = MakeBrighterFatterKernelTaskConfig
+    _DefaultName = "makeBrighterFatterKernel"
+
+    def __init__(self, *args, **kwargs):
+        """Constructor for the MakeBrighterFatterKernelTask."""
+        pipeBase.CmdLineTask.__init__(self, *args, **kwargs)
+        self.makeSubtask("isr")
+
+        self.debug = lsstDebug.Info(__name__)
+        if self.debug.enabled:
+            self.log.info("Running with debug enabled...")
+            # if we're displaying, test it works and save the displays for later
+            # it's worth testing here as displays are flaky and sometimes
+            # can't be contacted, and given processing takes a while
+            # it's a shame to fail late due to display issues
+            if self.debug.display:
+                try:
+                    afwDisp.setDefaultBackend(self.debug.displayBackend)
+                    afwDisp.Display.delAllDisplays()
+                    self.disp1 = afwDisp.Display(0, open=True)
+                    self.disp2 = afwDisp.Display(1, open=True)
+
+                    im = afwImage.ImageF(1, 1)
+                    im.array[:] = [[1]]
+                    self.disp1.mtv(im)
+                    self.disp1.erase()
+                except NameError:
+                    self.debug.display = False
+                    self.log.warn('Failed to setup/connect to display! Debug display has been disabled')
+
+        plt.interactive(False)  # stop windows popping up when plotting. When headless, use 'agg' backend too
+        self.checkIsrConfig()
+        self.config.validate()
+        self.config.freeze()
+
+    @classmethod
+    def _makeArgumentParser(cls):
+        """Augment argument parser for the MakeBrighterFatterKernelTask."""
+        parser = pipeBase.ArgumentParser(name=cls._DefaultName)
+        parser.add_argument("--visitPairs", nargs="*",
+                            help="Visit pairs to use. Each pair must be of the form INT,INT e.g. 123,456")
+        parser.add_id_argument("--id", datasetType="brighterFatterKernelNew",
+                               ContainerClass=BrighterFatterKernelTaskDataIdContainer,
+                               help="The ccds to use, e.g. --id ccd=0..100")
+        return parser
+
+    def checkIsrConfig(self):
+        """Check that appropriate ISR settings are being used
+        for brighter-fatter kernel calculation."""
+        mandatory = ['doAssembleCcd', 'doOverscan']  # raise if False
+        forbidden = ['doApplyGains', 'normalizeGains', 'doFlat', 'doFringe', 'doAddDistortionModel',
+                     'doBrighterFatter', 'doUseOpticsTransmission', 'doUseFilterTransmission',
+                     'doUseSensorTransmission', 'doUseAtmosphereTransmission', 'doGuider', 'doStrayLight',
+                     'doTweakFlat']  # raise if True
+        desirableTrue = ['doBias', 'doDark', 'doCrosstalk', 'doDefect', 'doLinearize']  # WARN if False
```

Comment:
I suggest making these three fields config parameters, so an obs package can plug in the appropriate variant ISR task and tweak the tests accordingly. (I thought they already were config fields, which explains an earlier confusion -- I had suggested that a help string point to the config fields which specify these tests).

Unfortunately the field name should probably include "isr", e.g. "isrMandatorySteps", "isrForbiddenSteps" and "isrDesirableSteps" -- with a possible future "isrUndesirableSteps".

If you make this change then I also strongly suggest failing if a specified ISR config field doesn't exist (e.g. e.g. by testing and raising -- or skipping the test if the resulting error message is clear enough). Otherwise you risk failing to test something important due to a spelling error. In fact that is another argument for making the tests more visible by making them part of the config.

Reply:
OK, I've moved all this to configs as you said. For the checking, what I've done is to remove the protecting blocks that were in there for the forbidden/mandatory catagories so that you get an error if they're not found, I think that should suffice. This made sense too as they're now overridable per-obs_package, so no reason to be forgiving if they're not there.

If you could just take a look and check that the way I've gone about this is OK with you I think that would be good, thanks.

Reply:
I just looked at the code. The new config parameters look fine (though I'm disappointed that one still needs to be hard-coded -- assembleCcd.doTrim; if you used dotted names for subtasks you could avoid that). The code I found using these is the `validateIsrConfig` method and it still logs a warning if a config parameter in one of these lists is missing -- rather than failing. So I'm not sure why you said "you get an error if they are not found". I'd prefer an error, but I think what you have will do.

Reply:
I think this is the last point remaining, but I've been struggling to understand here. What do you mean by "if you used dotted names for subtasks"? I tried to work it out but failed I'm afraid. Do you think you could give me a tiny example? I think regardless of whether it really needs changing I would learn something here.

Regarding the behaviour; mandatory/forbidden now raise, but desirable WARNs if wrong polarity, and logs an INFOs if the key is not found. I like this behaviour because it suggests to the user that they should configure the equivalent keys in the obs_package config, but this could easily be changed to a `raise` if not found if you think that's better (though I think it has to WARN if wrong, otherwise it's not "desirable", it's mandatory.) Does that sound good?

I'm totally happy to change this until you're happy though, so if you can confirm what you'd like to see instead I'll go ahead and make the changes (especially if I can move the hard-coded `isr.assembleCcd.doTrim` into the `mandatory` config).

Reply:
I just mean that the name in the config could be "assembleCcd.doTrim". Parsing that is a bit more work, as `getattr` and `setattr` cannot handle multiple levels in one call. However, a simple loop like the following would suffice (this is only lightly tested):
```
def getDottedAttr(obj, name):
    """Get an attribute that may be several levels deep, e.g. foo.bar.baz
    """
    if not name:
        raise ValueError("Name is blank")
    try:
        nameList = name.split(".")
        value = getattr(obj, nameList[0])
        for subName in nameList[1:]:
            value = getattr(value, subName)
        return value
    except Exception as e:
        raise AttributeError("Could not find {} in {}: {}".format(name, obj, e))
```

Reply:
And just to make it clear: I suggest using this function for all values you are trying to get, not just ones you know are dotted. The loop is not run at all if nameList has only one element, so it'll be quite efficient.